flang/lib/Lower/SymbolMap.h

```diff
@@ -33,84 +34,34 @@ namespace Fortran::lower {
 ///
 /// The lowering bridge needs to be able to record all four of these ssa-values
 /// in the lookup table to be able to correctly lower Fortran to FIR.
-struct SymIndex {
+struct SymbolBox {
```

Comment:
A symbol box is like any other tuple (box) of ssa-values, with the distinction it appears in the symbol table so the value and its property values can be looked up by name. All symbols that are valid on entry to the subprogram should be placed in the map. Even an unallocated variable should have a sentinel value bound to its symbol (e.g., undefined).

How the symbol table is managed is a separable problem. It is not solved in this PR.

I disagree with the final statement. A symbol bound in a context to a related set of ssa-values (say a heap address and various dynamic properties) must be maintained correctly as an aggregate and a SymbolBox is the structure for collecting that information.  Again the lifetime and scope of this box of information on an entity must be maintained in proper SSA form per the control-flow. The code to do that and manage the symbol table is a TODO.

Reply:
Ok, let's take an example:

```
subtroutine sub(x, y)
 real, allocatable :: x(:, :), y(:, :)
 if (foo(x)) then
   x = transpose(x)
 else
  call bar(x)
 end if
 if (allocated(x)) then
   y = x ! What's x shape here ?
 end if
end subroutine
```
Basically every statement that touches `x` may modify its shape and allocate/deallocate it (also when it escaped in calls). If at every point of lowering, we want the information inside the bridge to be correct, that would require:
1. catching all the places such information may change and reloading and reading back the descriptor to update the symbol table.
2. handling the structure flow, when two paths that modified an `x` merged, the symbol map must be updated with the value from the path that were taken....

Now, add parallelism into the equation and add the fact that one thread may modify a pointer/allocatable while the other threads are waiting somewhere else for it to be updated (in a well defined but bridge unfriendly way) and so on...
All that to say, I think it is very tricky to ensure that the bridge can safely update the SymbolMap with the correct shape + memory address information at the right time and it would be way safer to query the symbol descriptor every time (or at least once per statement) while lowering. An mlir pass could get rid of the useless reads after an variable allocation/deallocation + CFG analysis.

Reply:
OK. This is a discussion worth having.

OTOH, this PR isn't trying to solve any of these problems. I'm not sure this is the right place to discuss allocatables, etc. We're just trying to incrementally advance the ball out of our own goal area. ;)

This PR does not eliminate the runtime model of descriptors in any way. Those are still present and will still be used to model _runtime_ box values. There remains a distinction to draw between compile-time boxes in the bridge and runtime boxes, which will produce more object code, etc.